client/driver/docker_progress.go

```diff
@@ -0,0 +1,242 @@
+package driver
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/docker/docker/pkg/jsonmessage"
+)
+
+const (
+	// defaultPullActivityDeadline is the default value set in the imageProgressManager
+	// when newImageProgressManager is called
+	defaultPullActivityDeadline = 2 * time.Minute
+
+	// defaultImageProgressReportInterval is the default value set in the
+	// imageProgressManager when newImageProgressManager is called
+	defaultImageProgressReportInterval = 10 * time.Second
+)
+
+type layerProgress struct {
+	id           string
+	status       layerProgressStatus
+	currentBytes int64
+	totalBytes   int64
+}
+
+type layerProgressStatus int
+
+const (
+	layerProgressStatusUnknown layerProgressStatus = iota
+	layerProgressStatusStarting
+	layerProgressStatusWaiting
+	layerProgressStatusDownloading
+	layerProgressStatusVerifying
+	layerProgressStatusDownloaded
+	layerProgressStatusExtracting
+	layerProgressStatusComplete
+	layerProgressStatusExists
+)
+
+func lpsFromString(status string) layerProgressStatus {
+	switch status {
+	case "Pulling fs layer":
```

Comment:
Hmm, I'm thinking it might be a better idea to downcase these or compile regexes for the cases incase they change? The case variations make me worry this this may break with an update.